xbout/load.py

```diff
@@ -156,6 +156,98 @@ def open_boutdataset(datapath='./BOUT.dmp.*.nc', inputfilepath=None,
     return ds
 
 
+def collect(varname, xind=None, yind=None, zind=None, tind=None,
+            path=".", yguards=False, xguards=True, info=True, prefix="BOUT.dmp"):
+
+    from os.path import join
+
+    """
+
+    Extract the data pertaining to a specified variable in a BOUT++ data set
+
+
+    Parameters
+    ----------
+    varname : str
+        Name of the variable
+    xind, yind, zind, tind : int, slice or list of int, optional
+        Range of X, Y, Z or time indices to collect. Either a single
+        index to collect, a list containing [start, end] (inclusive
+        end), or a slice object (usual python indexing). Default is to
+        fetch all indices
+    path : str, optional
+        Path to data files (default: ".")
+    prefix : str, optional
+        File prefix (default: "BOUT.dmp")
+    yguards : bool, optional
+        Collect Y boundary guard cells? (default: False)
+    xguards : bool, optional
+        Collect X boundary guard cells? (default: True)
+        (Set to True to be consistent with the definition of nx)
+    info : bool, optional
+        Print information about collect? (default: True)
+
+    Notes
+    ----------
+    strict : This option found in boutdata.collect() is not present in this function
+             it is assumed that the varname given is correct, if variable does not exist
+             the function will fail
+    tind_auto : This option is not required when using _auto_open_mfboutdataset as an
+             automatic failure if datasets are different lengths is included
+
+    Returns
+    ----------
+    ds : numpy.ndarray
+
+    """
+
+    datapath = join(path, prefix + "*.nc")
+
+    ds = _auto_open_mfboutdataset(datapath, keep_xboundaries=xguards,
+                                  keep_yboundaries=yguards, info=info)
+
+    if varname not in ds:
+        raise KeyError("No variable, {} was found in {}.".format(varname, datapath))
+
+    dims = list(ds.dims)
+    inds = [tind, xind, yind, zind]
+
+    selection = {}
+
+    # Convert indexing values to an isel suitable format
+    for dim, ind in zip(dims, inds):
+
+        if isinstance(ind, int):
+            indexer = [ind]
+        elif isinstance(ind, list):
+            start, end = ind
+            indexer = slice(start, end)
```

Comment:
The old `collect()`'s indexing is a bit unorthodox. We probably need to add some tests of all possible types of `tind/xind/yind/zind` to check, if we really want to be backward compatible.

Here, I think when a list is passed, e.g. `xind = [start, end]`, for old `collect` the end should be included, so `slice(start, end+1)`. I think it's also allowed to have `xind = [start, end, step]` which would convert to `slice(start, end+1, step)`.

Reply:
Just did some quick tests. The `old_collect` function does in its docs say that if given a list e.g. `xind = [start, end]` that it is inclusive on the end, but I have test that exact case and it isn't inclusive. 

It will take `xind = [start, end, step]`, so I will write this into the new collect function along with tests to show that the indexing is not inclusive of the end.

Reply:
We've done some work on the slicing in `collect` recently (in the last year or so) so the answer may be slightly sensitive to the exact version, but for the current master of `BOUT-dev`, and I think also for the version in the `boutproject/boutdata` repo, a list like `xind = [start, end]` is inclusive. See https://github.com/boutproject/BOUT-dev/blob/712b7a1b9aeabe2e6c619cdf1d8d0d3dcbd26bf8/tools/pylib/boutdata/collect.py#L116-L117. The 3-element `xind = [start, stop, end]` form is different (currently exclusive of the `end` index) - I think this is a bug in `collect` rather than intentional.

Reply:
See https://github.com/boutproject/BOUT-dev/pull/1861

Reply:
Yes this was version sensitive I have updated my version, will update the `new_collect` function to match the current version of `boutdata`